Source/Core/DolphinQt2/CheatsManager.cpp

```diff
@@ -0,0 +1,624 @@
+          &CheatsManager::OnMatchContextMenu);
+  connect(m_watch_table, &QTableWidget::customContextMenuRequested, this,
+          &CheatsManager::OnWatchContextMenu);
+  connect(m_watch_table, &QTableWidget::itemChanged, this, &CheatsManager::OnWatchItemChanged);
+}
+
+void CheatsManager::OnWatchContextMenu()
+{
+  if (m_watch_table->selectedItems().isEmpty())
+    return;
+
+  QMenu* menu = new QMenu(this);
+
+  AddAction(menu, tr("Remove from Watch"), this, [this] {
+    auto* item = m_match_table->selectedItems()[0];
+
+    int index = item->data(INDEX_ROLE).toInt();
+
+    m_watch.erase(m_watch.begin() + index);
+
+    Update();
+  });
+
+  menu->addSeparator();
+
+  AddAction(menu, tr("Generate Action Replay Code"), this, &CheatsManager::GenerateARCode);
+
+  menu->exec(QCursor::pos());
+}
+
+void CheatsManager::OnMatchContextMenu()
+{
+  if (m_match_table->selectedItems().isEmpty())
+    return;
+
+  QMenu* menu = new QMenu(this);
+
+  AddAction(menu, tr("Add to Watch"), this, [this] {
+    auto* item = m_match_table->selectedItems()[0];
+
+    int index = item->data(INDEX_ROLE).toInt();
+
+    m_watch.push_back(m_results[index]);
+
+    Update();
+  });
+
+  menu->exec(QCursor::pos());
+}
+
+static ActionReplay::AREntry ResultToAREntry(Result result)
+{
+  u8 cmd;
+
+  switch (result.type)
+  {
+  case DataType::Byte:
+    cmd = 0x00;
+  case DataType::Short:
+    cmd = 0x02;
+  default:
+  case DataType::Int:
+    cmd = 0x04;
+  }
+
+  u32 address = result.address & 0xffffff;
+
+  return ActionReplay::AREntry(cmd << 24 | address, result.locked_value);
+}
+
+void CheatsManager::GenerateARCode()
+{
+  if (!m_ar_code)
+    return;
+
+  auto* item = m_match_table->selectedItems()[0];
+
+  int index = item->data(INDEX_ROLE).toInt();
+  ActionReplay::ARCode ar_code;
+
+  ar_code.active = true;
+  ar_code.user_defined = true;
+  ar_code.name = tr("Generated by search (Address %1)")
+                     .arg(m_watch[index].address, 8, 16, QLatin1Char('0'))
+                     .toStdString();
+
+  ar_code.ops.push_back(ResultToAREntry(m_watch[index]));
+
+  m_ar_code->AddCode(ar_code);
+}
+
+void CheatsManager::OnWatchItemChanged(QTableWidgetItem* item)
+{
+  if (m_updating)
+    return;
+
+  int index = item->data(INDEX_ROLE).toInt();
+  int column = item->data(COLUMN_ROLE).toInt();
+
+  switch (column)
+  {
+  case 0:
+    m_watch[index].name = item->text();
+  case 3:
+    m_watch[index].locked = item->checkState() == Qt::Checked;
+    break;
+  case 4:
+  {
+    const auto text = item->text();
+    u64 value = 0;
+
+    switch (static_cast<DataType>(m_match_length->currentIndex()))
+    {
+    case DataType::Byte:
+      value = text.toUShort(nullptr, 16) & 0xFF;
+    case DataType::Short:
+      value = text.toUShort(nullptr, 16);
+    case DataType::Int:
+      value = text.toUInt(nullptr, 16);
+    case DataType::Float:
+    {
+      float f = text.toFloat();
+      value = *reinterpret_cast<u32*>(&f);
```

Comment:
This is undefined behavior. A memcpy should be done instead.

Reply:
Why though? Isn't f garantuued to be aligned properly?

Reply:
They're completely unrelated types. It breaks the strict aliasing rule. The only types you can cast to and have this guaranteed to work are `std::byte`, `char` and `unsigned char`. e.g. in the standard (emphasis added by me):

> If a program attempts to access the stored value of an object through a glvalue of **other than one of the following types** the behavior is undefined:
>
> - the dynamic type of the object,
> - a cv-qualified version of the dynamic type of the object,
> - a type similar (as defined in 7.5) to the dynamic type of the object,
> - a type that is the signed or unsigned type corresponding to the dynamic type of the object,
> - a type that is the signed or unsigned type corresponding to a cv-qualified version of the dynamic type of the object,
> - an aggregate or union type that includes one of the aforementioned types among its elements or nonstatic data members (including, recursively, an element or non-static data member of a subaggregate or contained union),
> - a type that is a (possibly cv-qualified) base class type of the dynamic type of the object,
> - a `char`, `unsigned char`, or `std::byte` type.

in other words (with the relevant bullet points translated to somewhat English; the union and base class bullet points are inherited from C and don't apply to C++, see [this](http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#2051) if you want more info on that):

> Given the following conversion:
>
> ```cpp
> SRC object = ...;
> DST dst_object = *reinterpret_cast<DST*>(&object)
> ```
>
> The behavior is undefined unless one of the following is true:
>
> - DST is the same type as SRC
> - DST is a const-qualified, volatile-qualified (or both) variant of SRC
> - DST is a signed or unsigned (and possibly const-qualified, volatile-qualified, or both) variant of SRC
> - DST is either`char`, `unsigned char` or `std::byte`

`float` satisfies none of these for conversions to a `u32`

Reply:
Weird. But a nice explanation anyway. Thanks!